crates/ruff_python_ast/generate.py

```diff
@@ -693,6 +746,158 @@ def write_anynoderef(out: list[str], ast: Ast) -> None:
         """)
 
 
+# ------------------------------------------------------------------------------
+# AnyRootNodeRef
+
+
+def write_root_anynoderef(out: list[str], ast: Ast) -> None:
+    """
+    Create the AnyRootNodeRef type.
+
+    ```rust
+    pub enum AnyRootNodeRef<'a> {
+        ...
+        TypeParam(&'a TypeParam),
+        ...
+    }
+    ```
+
+    Also creates:
+    - `impl<'a> From<&'a TypeParam> for AnyRootNodeRef<'a>`
+    - `impl<'a> TryFrom<AnyRootNodeRef<'a>> for &'a TypeParam`
+    - `impl<'a> TryFrom<AnyRootNodeRef<'a>> for &'a TypeParamVarTuple`
+    - `impl Ranged for AnyRootNodeRef<'_>`
+    - `impl HasNodeIndex for AnyRootNodeRef<'_>`
+    - `fn AnyRootNodeRef::visit_source_order(self, visitor &mut impl SourceOrderVisitor)`
+    """
+
+    out.append("""
+    /// An enumeration of all AST nodes.
+    ///
+    /// Unlike `AnyNodeRef`, this type does not flatten nested enums, so its
+    /// variants only consist of the "root" AST node types. This is useful as
+    /// it exposes references to the original enums, not just references to their
+    /// inner values.
```

Comment:
We can add an example here to explain what does "root" mean like we can say that this will include the top level `Stmt` enum but will not include the individual variants of the `Stmt` enum.